ttnn/cpp/ttnn/device_operation.hpp

```diff
@@ -254,99 +262,188 @@ inline void log_operation(
 
     auto& program_cache = device->program_cache;
 
-    if (program_cache.is_enabled()) {
-
-        auto program_hash = compute_program_hash<device_operation_t>(operation_attributes, tensor_args);
-        auto program_cache_hit = program_cache.contains(program_hash);
+    auto program_hash = 0;
+    bool program_cache_hit = false;
 
-        log_operation<device_operation_t>(operation_attributes, tensor_args, program_hash, program_cache_hit);
+    auto is_program_cache_enabled = program_cache.is_enabled();
+    if (is_program_cache_enabled) {
+        program_hash = compute_program_hash<device_operation_t>(operation_attributes, tensor_args);
+        program_cache_hit = program_cache.contains(program_hash);
+    }
 
-        tt::stl::reflection::visit_object_of_type<Tensor>(CheckDeviceBufferIsAllocated{}, tensor_args);
+    log_device_operation_start<device_operation_t>(
+            operation_id,
+            primitive_operation_id,
+            device->id(),
+            operation_attributes,
+            tensor_args,
+            program_hash,
+            program_cache_hit
+        );
 
-        if (program_cache_hit) {
-            ZoneScopedN("Validate on Program Cache Hit");
-            device_operation_t::validate_on_program_cache_hit(operation_attributes, tensor_args);
-        } else {
-            ZoneScopedN("Validate on Program Cache Miss");
-            device_operation_t::validate_on_program_cache_miss(operation_attributes, tensor_args);
-        }
+    tt::stl::reflection::visit_object_of_type<Tensor>(CheckDeviceBufferIsAllocated{}, tensor_args);
 
-        auto& program = create_or_get_program_from_cache<device_operation_t>(
-            program_cache, program_cache_hit, program_hash, operation_attributes, tensor_args, tensor_return_value);
+    if (program_cache_hit) {
+        ZoneScopedN("Validate on Program Cache Hit");
+        device_operation_t::validate_on_program_cache_hit(operation_attributes, tensor_args);
+    } else {
+        ZoneScopedN("Validate on Program Cache Miss");
+        device_operation_t::validate_on_program_cache_miss(operation_attributes, tensor_args);
+    }
 
+    const auto enqueue_or_launch_program = [=](auto& program) {
         if (USE_FAST_DISPATCH) {
             ZoneScopedN("EnqueueProgram");
             auto& queue = device->command_queue(cq_id);
             tt::tt_metal::EnqueueProgram(queue, program, false);
         } else {
             ZoneScopedN("LaunchProgram");
-            ::detail::LaunchProgram(device, program);
+            tt::tt_metal::detail::LaunchProgram(device, program);
         }
+    };
+
+    if (is_program_cache_enabled) {
+        auto& program = create_or_get_program_from_cache<device_operation_t>(
+            program_cache, program_cache_hit, program_hash, operation_attributes, tensor_args, tensor_return_value);
+
+        enqueue_or_launch_program(program);
+
+        log_device_operation_end<device_operation_t>(
+            operation_id,
```

Comment:
Is operation_id unique per op type and primitive_operation_id unique per op instance?

Reply:
@mo-tenstorrent yes. It's used the same way as the id you added in the op_profiler